diff --git a/composer.json b/composer.json
index d02c302..86a53de 100755
--- a/composer.json
+++ b/composer.json
@@ -13,9 +13,7 @@
   "replace": {
     "islandora/islandora_mirador": "self.version"
   },
-  "require": {
-    "drupal/islandora": "^2"
-  },
+  "require": {},
   "conflict": {
     "drupal/core": "<=8",
     "islandora/islandora_defaults": "<3"
diff --git a/src/Plugin/Field/FieldFormatter/MiradorImageFormatter.php b/src/Plugin/Field/FieldFormatter/MiradorImageFormatter.php
index 3f1250a..c13c4f6 100644
--- a/src/Plugin/Field/FieldFormatter/MiradorImageFormatter.php
+++ b/src/Plugin/Field/FieldFormatter/MiradorImageFormatter.php
@@ -10,8 +10,8 @@ use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
 use Drupal\Core\Routing\RouteMatchInterface;
 use Drupal\Core\Utility\Token;
 use Drupal\image\Plugin\Field\FieldFormatter\ImageFormatterBase;
-use Drupal\islandora\IslandoraUtils;
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
 
 /**
  * Mirador FieldFormatter plugin.
@@ -51,11 +51,11 @@ class MiradorImageFormatter extends ImageFormatterBase implements ContainerFacto
   protected $routeMatch;
 
   /**
-   * Islandora utility functions.
+   * The entity type manager.
    *
-   * @var \Drupal\islandora\IslandoraUtils
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
    */
-  protected $utils;
+  protected $entityTypeManager;
 
   /**
    * Constructs a StringFormatter instance.
@@ -80,15 +80,15 @@ class MiradorImageFormatter extends ImageFormatterBase implements ContainerFacto
    *   The token service.
    * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
    *   The route match.
-   * @param \Drupal\islandora\IslandoraUtils $utils
-   *   Islandora utils.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
    */
-  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, $label, $view_mode, array $third_party_settings, ConfigFactoryInterface $config_factory, Token $token, RouteMatchInterface $route_match, IslandoraUtils $utils) {
+  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, $label, $view_mode, array $third_party_settings, ConfigFactoryInterface $config_factory, Token $token, RouteMatchInterface $route_match, EntityTypeManagerInterface $entity_type_manager) {
     parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $label, $view_mode, $third_party_settings, $config_factory);
     $this->token = $token;
     $this->routeMatch = $route_match;
     $this->configFactory = $config_factory;
-    $this->utils = $utils;
+    $this->entityTypeManager = $entity_type_manager;
   }
 
   /**
@@ -106,7 +106,7 @@ class MiradorImageFormatter extends ImageFormatterBase implements ContainerFacto
       $container->get('config.factory'),
       $container->get('token'),
       $container->get('current_route_match'),
-      $container->get('islandora.utils')
+      $container->get('entity_type.manager')
     );
   }
 
@@ -126,7 +126,7 @@ class MiradorImageFormatter extends ImageFormatterBase implements ContainerFacto
     foreach ($files as $file) {
       $cache_meta = CacheableMetadata::createFromObject($file);
 
-      $medias = $this->utils->getReferencingMedia($file->id());
+      $medias = $this->getReferencingMedia($file->id());
       $first_media = array_values($medias)[0];
       $cache_meta->addCacheableDependency($first_media);
       $node = $first_media->get('field_media_of')->entity;
@@ -146,4 +146,66 @@ class MiradorImageFormatter extends ImageFormatterBase implements ContainerFacto
     return $elements;
   }
 
+  /**
+   * Gets Media that reference a File.
+   *
+   * @param int $fid
+   *   File id.
+   *
+   * @return \Drupal\media\MediaInterface[]
+   *   Array of media.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   *   Calling getStorage() throws if the entity type doesn't exist.
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   *   Calling getStorage() throws if the storage handler couldn't be loaded.
+   */
+  public function getReferencingMedia($fid) {
+    // Get media fields that reference files.
+    $fields = $this->getReferencingFields('media', 'file');
+
+    // Process field names, stripping off 'media.' and appending 'target_id'.
+    $conditions = array_map(
+      function ($field) {
+        return ltrim($field, 'media.') . '.target_id';
+      },
+      $fields
+    );
+
+    // Query for media that reference this file.
+    $query = $this->entityTypeManager->getStorage('media')->getQuery();
+    $query->accessCheck(TRUE);
+    $group = $query->orConditionGroup();
+    foreach ($conditions as $condition) {
+      $group->condition($condition, $fid);
+    }
+    $query->condition($group);
+
+    return $this->entityTypeManager->getStorage('media')
+      ->loadMultiple($query->execute());
+  }
+
+  /**
+   * Get the fields on an entity of $entity_type that reference a $target_type.
+   *
+   * @param string $entity_type
+   *   Type of entity to search for.
+   * @param string $target_type
+   *   Type of entity the field references.
+   *
+   * @return array
+   *   Array of fields.
+   */
+  public function getReferencingFields($entity_type, $target_type) {
+    $fields = $this->entityTypeManager->getStorage('field_storage_config')
+      ->getQuery()
+      ->condition('entity_type', $entity_type)
+      ->condition('settings.target_type', $target_type)
+      ->execute();
+    if (!is_array($fields)) {
+      $fields = [$fields];
+    }
+    return $fields;
+  }
+
 }
diff --git a/islandora_mirador.module b/islandora_mirador.module
index c1a8bbe..ca18d40 100644
--- a/islandora_mirador.module
+++ b/islandora_mirador.module
@@ -35,6 +35,19 @@ function islandora_mirador_theme() {
 function template_preprocess_mirador(&$variables) {
   $variables['mirador_view_id'] = Html::getUniqueId($variables['mirador_view_id']);
 
+  // For dsu-utsc.
+  if (!empty(\Drupal::hasService('jwt.authentication.jwt'))) {
+    $jwtService = \Drupal::service('jwt.authentication.jwt');
+    $token = $jwtService->generateToken();
+
+    // Attach the valid token as a Drupal setting
+    $variables['#attached']['drupalSettings']['token'] = $token;
+
+    // Add cache metadata to limit the cache lifetime
+    $variables['#cache']['max-age'] = 3600; // 1 hour
+    $variables['#cache']['tags'][] = 'jwt_token'; // Add a cache tag to invalidate the cache when the token changes
+  }
+
   /**
    * @var \Drupal\islandora_mirador\IslandoraMiradorPluginManager
    */
diff --git a/islandora_mirador.routing.yml b/islandora_mirador.routing.yml
index 9d27632..c377ef8 100644
--- a/islandora_mirador.routing.yml
+++ b/islandora_mirador.routing.yml
@@ -5,3 +5,9 @@ islandora_mirador.miradorconfig:
     _title: 'Mirador Settings'
   requirements:
     _permission: 'administer site configuration'
+islandora_mirador.service_worker:
+  path: '/islandora_mirador_service_worker'
+  defaults: 
+    _controller: '\Drupal\islandora_mirador\Controller\ServiceWorkerController::serve' 
+  requirements: 
+    _permission: 'access content'
diff --git a/js/mirador_viewer.js b/js/mirador_viewer.js
index c520e40..2f4f33d 100644
--- a/js/mirador_viewer.js
+++ b/js/mirador_viewer.js
@@ -5,33 +5,96 @@
  * Displays Mirador viewer.
  */
 (function (Drupal, once) {
-    'use strict';
+  'use strict';
 
-    /**
-     * Initialize the Mirador Viewer.
-     */
-    Drupal.behaviors.Mirador = {
-        attach: function (context, settings) {
-            Drupal.IslandoraMirador = Drupal.IslandoraMirador || {}
-            Drupal.IslandoraMirador.instances = Drupal.IslandoraMirador.instances || {}
-            Object.entries(settings.mirador.viewers).forEach(entry => {
-              const [base, values] = entry;
-              once('mirador-viewer', base, context).forEach(() =>
-                // save the mirador instance so other modules can interact
-                // with the store/actions at e.g. Drupal.IslandoraMirador.instances["#mirador-xyz"].store
-                Drupal.IslandoraMirador.instances[base] = Mirador.viewer(values, window.miradorPlugins || {})
-              );
-            });
-        },
-        detach: function (context, settings) {
-            Object.entries(settings.mirador.viewers).forEach(entry => {
-              const [base, ] = entry;
-              const removed = once.remove('mirador-viewer', base, context);
-              if (removed.length > 0) {
-                delete Drupal.IslandoraMirador.instances[base];
+  /**
+   * Initialize the Mirador Viewer.
+   */
+  Drupal.behaviors.Mirador = {
+      attach: function (context, settings) {
+          Drupal.IslandoraMirador = Drupal.IslandoraMirador || {}
+          Drupal.IslandoraMirador.instances = Drupal.IslandoraMirador.instances || {}
+          Object.entries(settings.mirador.viewers).forEach(entry => {
+            const [base, values] = entry;
+            once('mirador-viewer', base, context, settings).forEach(() => {
+              if (settings.token !== undefined) {
+                values["resourceHeaders"] = {
+                  'Authorization': 'Bearer '+ settings.token,
+                  'token': settings.token
+                };
+                values["requestPipeline"] = [
+                  (url, options) => ({  ...options, headers: {
+                    "Accept": 'application/ld+json;profile="http://iiif.io/api/presentation/3/context.json"',
+                    'Authorization': 'Bearer '+ settings.token,
+                    'token': settings.token
+                  }})
+                ];
+                values["osdConfig"] = {
+                  "loadTilesWithAjax": true,
+                  "ajaxHeaders": {
+                    'Authorization': 'Bearer '+ settings.token,
+                    'token': settings.token
+                  }
+                };
+                values["requests"] = {
+                  preprocessors: [ // Functions that receive HTTP requests and manipulate them (e.g. to add headers)
+                    // rewrite all info.json requests to add the text/json request header
+                    (url, options) => (url.match('info.json') && { ...options, headers: {
+                      'Authorization': 'Bearer '+ settings.token,
+                      'token': settings.token
+                    }})
+                  ],
+                };
               }
+              Drupal.IslandoraMirador.instances[base] = Mirador.viewer(values, window.miradorPlugins || {})
             });
-        }
-    };
+          });
+
+          if (settings.token !== undefined) {
+              if ('serviceWorker' in navigator) {
+                // Unregister the existing service worker
+                navigator.serviceWorker.getRegistrations()
+                  .then(registrations => {
+                    registrations.forEach(registration => {
+                      var scope_url = window.location.protocol + "//" + window.location.hostname + "/";
+                      if (registration.scope === scope_url
+                            && registration.waiting != null && registration.waiting.scriptURL.includes('islandora_mirador_service_worker')) {
+                         registration.unregister().then((boolean) => {
+                            // if boolean = true, unregister is successful
+                            if (boolean == true)
+                               console.log("unregister successfully ");
+                         });
+                      }
+                    });
+                  })
+                  .catch(err => {
+                    console.error('Error unregistering service worker:', err);
+                  })
+                  .finally(() => {
+                    // Register the new service worker with the token
+                    window.addEventListener('load', () => {
+                      navigator.serviceWorker
+                        .register('/islandora_mirador_service_worker?token=' + settings.token, { scope: '/' })
+                        .then(registration => {
+                          console.log('ServiceWorker registration successful with scope: ', registration.scope);
+                        })
+                        .catch(err => {
+                          console.log('ServiceWorker registration failed: ', err);
+                        });
+                    });
+                  });
+              }
+           }
+      },
+      detach: function (context, settings) {
+          Object.entries(settings.mirador.viewers).forEach(entry => {
+            const [base, ] = entry;
+            const removed = once.remove('mirador-viewer', base, context);
+            if (removed.length > 0) {
+              delete Drupal.IslandoraMirador.instances[base];
+            }
+          });
+      }
+  };
 
 })(Drupal, once);
diff --git a/js/service_worker.js b/js/service_worker.js
new file mode 100644
index 0000000..0804b3c
--- /dev/null
+++ b/js/service_worker.js
@@ -0,0 +1,20 @@
+self.addEventListener('activate', function (event) {
+  return self.clients.claim();
+});
+
+self.addEventListener('fetch', function (event) {
+  if (event.request.destination === "image" && new URL(event.request.url).pathname.startsWith('/iiif/') && new URL(location).searchParams.has('token')) {
+    var token = new URL(location).searchParams.get('token');
+    event.respondWith(
+      fetch(event.request, {
+        headers: {
+          'Authorization': 'Bearer ' + token,
+          'token': token
+        },
+        mode: "cors",
+        credentials: "omit"
+        //credentials: "include"
+      })
+    );
+  }
+});
diff --git a/src/Controller/ServiceWorkerController.php b/src/Controller/ServiceWorkerController.php
new file mode 100644
index 0000000..fb677b6
--- /dev/null
+++ b/src/Controller/ServiceWorkerController.php
@@ -0,0 +1,58 @@
+<?php
+
+namespace Drupal\islandora_mirador\Controller;
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Extension\ExtensionPathResolver;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * A controller for the Service Worker.
+ */
+class ServiceWorkerController extends ControllerBase {
+
+  /**
+   * The extension path resolver service.
+   *
+   * @var \Drupal\Core\Extension\ExtensionPathResolver
+   */
+  protected $extensionPathResolver;
+
+  /**
+   * Constructs a ServiceWorkerController object.
+   *
+   * @param \Drupal\Core\Extension\ExtensionPathResolver $extension_path_resolver
+   *   The extension path resolver service.
+   */
+  public function __construct(ExtensionPathResolver $extension_path_resolver) {
+    $this->extensionPathResolver = $extension_path_resolver;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('extension.path.resolver')
+    );
+  }
+
+  /**
+   * Adds headers to the HTTP response.
+   */
+  public function serve(Request $request) {
+    $file_str = $this->extensionPathResolver->getPath('module', 'islandora_mirador') . '/js/service_worker.js';
+    if (file_exists($file_str)) {
+      $response = new BinaryFileResponse($file_str, 200);
+      $response->headers->set('Content-Type', 'application/javascript');
+      // Allow same origin service worker.
+      $response->headers->set('Service-Worker-Allowed', '/');
+      return $response;
+    }
+    throw new NotFoundHttpException();
+  }
+
+}
